lobbyboy/socket_handle.py

```diff
@@ -0,0 +1,249 @@
+        channel_fd = self.channel.fileno()
+        master_fd = server.master_fd
+        while proxy_subprocess.poll() is None:
+            r, *_ = select.select([master_fd, channel_fd], [], [], 0.1)
+            if master_fd in r:
+                send_to_channel(self.channel, os.read(master_fd, 10240).decode(), suffix="")
+            elif channel_fd in r:
+                os.write(master_fd, self.channel.recv(10240))
+
+    def cleanup(self, t: Transport = None, meta: LBServerMeta = None, check_destroy: bool = False):
+        if t and meta:
+            self.remove_server_session(t, meta.server_name)
+            if check_destroy:
+                self.destroy_server_if_needed(meta)
+
+        if self.channel:
+            self.channel.shutdown(0)
+        if t:
+            t.close()
+
+    def destroy_server_if_needed(self, server: LBServerMeta):
+        provider = self.providers[server.provider_name]
+        need_destroy, reason = self.killer.need_destroy(provider, server)
+        send_to_channel(self.channel, f"LobbyBoy: This server {reason}.")
+        if not need_destroy:
+            return
+
+        send_to_channel(self.channel, f"LobbyBoy: I will destroy {server.server_name}({server.server_host}) now!")
+        self.killer.destroy(provider, server, self.channel)
+        send_to_channel(
+            self.channel,
+            f"LobbyBoy: Server {server.server_name}({server.server_host}) has been destroyed.",
+        )
+
+    @staticmethod
+    def remove_server_session(transport: Transport, server_name: str):
+        peer_name = transport.getpeername()
+        with active_session_lock:
+            sessions = active_session.get(server_name)
+            if sessions:
+                active_session[server_name] = list(filter(lambda x: x.getpeername() != peer_name, sessions))
+
+    def run(self):
+        logger.info(
+            f"start new thread "
+            f"handle {self.socket_client}, "
+            f"address: {self.client_address}, "
+            f"my thread id={threading.get_ident()}"
+        )
+        t = Transport(self.socket_client, gss_kex=DoGSSAPIKeyExchange)
+        try:
+            t.set_gss_host(socket.getfqdn())
+            server = self.prepare_server(t)
+            if not (server and self.channel):
+                self.cleanup(t)
+                return
+
+            send_to_channel(self.channel, f"Welcome to LobbyBoy {__version__}!")
+            lb_server, proxy_subprocess = self.prepare_shell_env(server, t)
+            if not (proxy_subprocess and lb_server):
+                logger.error("failed to create proxy subprocess or lb_server")
+                self.cleanup(t, meta=lb_server)
+                return
+
+            self.user_using(server, proxy_subprocess)
+            send_to_channel(
+                self.channel,
+                f"LobbyBoy: SSH to remote server {lb_server.server_name} closed.",
+            )
+            self.cleanup(t, meta=server, check_destroy=True)
```

Comment:
the `server` here is a `lobbyboy.server.Server`, not a `LBServerMeta`

Reply:
Yes, let me fix now